chore: add support for one-way websockets to backend

buenos-nachos opened the pull request:
Closes https://github.com/coder/coder/issues/16775
Part 1/2, with https://github.com/coder/coder/pull/16855 handling the frontend changes

## Changes made
- Added `OneWayWebSocket` function that establishes WebSocket connections that don't allow client-to-server communication
- Added tests for the new function
- Updated API endpoints to make new WS-based endpoints, and mark previous SSE-based endpoints as deprecated
- Updated existing SSE handlers to use the same core logic as the new WS handlers

buenos-nachos committed:
chore: add support for one-way websockets to backend

buenos-nachos committed:
chore: make gen

Parkreiner commented:
@Emyrk Just pinging you in case you didn't see this. Let me know if you're busy, and I can find someone else to review this

f0ssel commented on coderd/httpapi/httpapi_test.go:
```diff
@@ -155,3 +160,263 @@ func TestWebsocketCloseMsg(t *testing.T) {
+	return w(b)
+}
+
+func TestOneWayWebSocket(t *testing.T) {
```
Because we also touch the SSE Sender, could we include some basic coverage for that if there isn't already?

f0ssel reviewed (commented):
Overall this looks great, the tests are really good coverage and the shared implementation worked out nice. 

I suggested some naming changes, mainly because we don't use "callback" very much in Go and like to use "-er" words to describe capabilities.

buenos-nachos committed:
Merge branch 'main' into mes/one-way-ws-01

buenos-nachos committed:
fix: apply feedback

buenos-nachos committed:
wip: commit progress on tests

buenos-nachos committed:
wip: commit more test progress

buenos-nachos committed:
chore: get all tests passing

Parkreiner commented:
@f0ssel Forgot to ping you yesterday, but I applied all your feedback and added tests for the SSE code

buenos-nachos committed:
Merge branch 'main' into mes/one-way-ws-01

buenos-nachos committed:
Merge branch 'main' into mes/one-way-ws-01

buenos-nachos committed:
chore: add support for one-way WebSockets to UI (#16855)

Closes https://github.com/coder/coder/issues/16777

## Changes made
- Added `OneWayWebSocket` utility class to help enforce one-way
communication from the server to the client
- Updated all client client code to use the new WebSocket-based
endpoints made to replace the current SSE-based endpoints
- Updated WebSocket event handlers to be aware of new protocols
- Refactored existing `useEffect` calls and removed some synchronization
bugs
- Removed dependencies and types for dealing with SSEs
- Addressed some minor Biome warnings

buenos-nachos committed:
Merge branch 'main' into mes/one-way-ws-01
